src/postgres_deparse.c

```diff
@@ -4902,14 +4904,32 @@ static void deparseCreateFunctionStmt(StringInfo str, CreateFunctionStmt *create
 		else
 		{
 			appendStringInfoString(str, "BEGIN ATOMIC ");
-			deparseExprList(str, castNode(List, create_function_stmt->sql_body));
+			if IsA(create_function_stmt->sql_body, List) {
+				foreach(lc, castNode(List, create_function_stmt->sql_body))
+				{
+					deparseTList(str, castNode(List, lfirst(lc)));
+				}
+			} else {
+				deparseExprList(str, castNode(List, create_function_stmt->sql_body));
+			}
```

Comment:
It actually looks from my reading of `gram.y` that this is always either a single-item list that contains NULL, or a single-item List that contains another List (https://github.com/postgres/postgres/blob/master/src/backend/parser/gram.y#L8605)

Re-reading the existing code, I don't know why I previously called `deparseExprList` here, but it seems wrong. Also I realize that the list items can either be a Stmt, or a ReturnStmt.

How about we change it like this:

```c
			if (linitial(create_function_stmt->sql_body) != NULL)
			{
				List *body_stmt_list = castNode(List, linitial(create_function_stmt->sql_body));
				foreach(lc, body_stmt_list)
				{
					if (IsA(lfirst(lc), ReturnStmt))
					{
						deparseReturnStmt(str, lfirst_node(ReturnStmt, lc));
					}
					else
					{
						deparseStmt(str, lfirst(lc));
						appendStringInfoString(str, "; ");
					}
				}
			}
```

(I think its good to keep this specific, since lists show up in other places, and so I don't think we should introduce a deparseTList that assumes `;` separators).

Reply:
Thanks for the suggestions. This fixes are seem to be more useful.